spec/src/main/asciidoc/Structure and format.adoc

```diff
@@ -0,0 +1,81 @@
+== Structure and format
+
+=== Terms
+
+.Terms and definitions
+Configuration:: An optional _top-level value_, with a collection of zero or more nested _configuration entries_ (organized by _configuration key_).
+  Also informally known as a "configuration tree node" or "configuration node".
+  A configuration with no value and no nested entries is considered to be an _empty configuration_.
+
+Configuration entry:: An _configuration key_ and _configuration value_ combination.
+
+Configuration key:: A non-empty string which uniquely identifies a configuration entry within a single nesting level of a configuration.
+  A configuration key *must not* contain embedded newlines or other control characters.
+  A configuration key string *must* be normalized using https://www.unicode.org/reports/tr15/#Norm_Forms)[Unicode Normalization Form C (NFC)].
+
+Configuration path:: A sequence of _configuration keys_ which uniquely identify a descendant _configuration entry_ within a configuration, at an arbitrary level of nesting.
+  Each configuration key within the sequence identifies one level of nested configuration.
+
+Configuration source:: A provider of configuration data in _raw value_ form. [TODO: high level structure of config sources, ordering etc]
+
+Configuration value:: An object which is stored within a _configuration_ and is accessible at the _configuration path_ which corresponds to that _configuration_.
+
+Empty configuration:: A configuration which has no _top-level value_ and zero nested _configuration entries_.
+
+Raw value:: A _configuration value_ in its uncoverted (string) form.
+
+Root:: The _configuration_ which is identified by a empty (zero-length) _configuration path_.
+  Every _configuration_ is its own _root_ when considered in isolation.
```

Comment:
Can configurations be nested?  If a configuration can be nested, then within its containing configuration it must have a configuration key, right, so you can draw a path to it? and a value?  Doesn't that make it a configuration entry?  Can these two terms be consolidated?

Reply:
The essential characteristic of the tree structure is that every tree level is a nested configuration.  I'll look into consolidating the terms and see if it can be done cogently.